keiko-sql/src/main/kotlin/com/netflix/spinnaker/q/sql/SqlQueue.kt

```diff
@@ -0,0 +1,743 @@
+    var missing = false
+
+    withRetry(RetryCategory.WRITE) {
+      jooq.transaction { config ->
+        val txn = DSL.using(config)
+
+        val queueRows = txn.select(fingerprintField)
+          .from(queueTable)
+          .where(fingerprintField.eq(fingerprint))
+          .limit(1)
+          .execute()
+
+        if (queueRows == 0) {
+          val unackedRows = txn.select(fingerprintField)
+            .from(unackedTable)
+            .where(fingerprintField.eq(fingerprint))
+            .limit(1)
+            .execute()
+
+          if (unackedRows == 0) {
+            missing = true
+          }
+        }
+      }
+    }
+
+    if (missing) {
+      log.debug(
+        "Pushing ensured message onto queue as it does not exist in queue or unacked tables"
+      )
+      push(message, delay)
+    }
+  }
+
+  private fun expireStaleLocks() {
+    val now = clock.instant().toEpochMilli()
+    val minMs = now.minus(TimeUnit.SECONDS.toMillis(lockTtlSeconds.toLong()))
+    val minUlid = ULID().nextValue(minMs).toString()
+
+    val rs = withRetry(RetryCategory.READ) {
+      jooq.select(idField, fingerprintField, deliveryField, lockedField)
+        .from(queueTable)
+        .where(
+          idField.lt(minUlid),
+          lockedField.ne("0")
+        )
+        .fetch()
+        .intoResultSet()
+    }
+
+    var ulid = ULID().nextValue()
+
+    while (rs.next()) {
+      val id = rs.getString("id")
+      val lock = rs.getString("locked")
+      val fingerprint = rs.getString("fingerprint")
+      val lockMatch = lockedAtRegex.find(lock)
+
+      if (lockMatch != null && lockMatch.groupValues.size > 1) {
+        if (lockMatch.groupValues[1].toLong() > minMs) {
+          /* Not time yet */
+          continue
+        }
+      } else {
+        log.error("Failed parsing lockedAt time for message id: $id, " +
+          "fingerprint: $fingerprint, lock: $lock, releasing")
+      }
+
+      /**
+       * No retries inside this transaction, transient failures here will be cleared up at
+       * the next retry() interval.
+       */
+      jooq.transaction { config ->
+        val txn = DSL.using(config)
+
+        val deleted = txn.delete(queueTable)
+          .where(idField.eq(id), lockedField.eq(lock))
+          .execute()
+
+        if (deleted == 1) {
+          log.info("releasing stale lock for fingerprint: $fingerprint")
+          ulid = ULID().nextMonotonicValue(ulid)
+
+          /**
+           * Re-insert with a fresh ulid and for immediate delivery
+           */
+          txn.insertInto(queueTable)
+            .set(idField, ulid.toString())
+            .set(fingerprintField, fingerprint)
+            .set(deliveryField, now)
+            .set(lockedField, "0")
+            .execute()
+        }
+      }
+    }
+  }
+
+  /**
+   * Differs from other [Queue.retry] implementations, as unacked messages are requeued for
+   * delivery at now + [lockTtlSeconds] instead of immediately.
+   */
```

Comment:
I like this departure, but why?

Reply:
I made this departure after pulling `jooq.update(messagesTable)` outside of the txn that moves a message from unacked back to the queue in order to minimize lock contention (comment at L611). If requeued for immediate delivery, there's a greater chance that the new `ackAttemptsAttribute` value will be lost. `lockTtlSeconds` is assumed to be a value within which all necessary db operations are safely completed for the system to be operational.